plugin/src/main/scala/org/scalafmt/sbt/ScalafmtPlugin.scala

```diff
@@ -146,135 +163,167 @@ object ScalafmtPlugin extends AutoPlugin {
   }
 
   private def checkSources(
-      cacheDirectory: File,
+      cacheStoreFactory: CacheStoreFactory,
       sources: Seq[File],
       config: Path,
       log: Logger,
       writer: PrintWriter
-  ): Boolean = {
-    cached(cacheDirectory, FilesInfo.lastModified, config) { cacheMisses =>
-      val changed = cacheMisses.filter(_.exists)
-      if (changed.size > 0) {
-        log.info(s"Checking ${changed.size} Scala sources...")
-        checkSources(changed.toSeq, config, log, writer)
-      } else {
-        true
-      }
-    }(sources.toSet).getOrElse(true)
+  ): ScalafmtAnalysis = {
+    trackSourcesAndConfig(cacheStoreFactory, sources, config) {
+      (outDiff, configChanged, prev) =>
+        log.debug(outDiff.toString)
+        val updatedOrAdded = outDiff.modified & outDiff.checked
+        val filesToCheck: Set[File] =
+          if (configChanged) sources.toSet
+          else updatedOrAdded
+        val prevFailed: Set[File] =
+          if (configChanged) Set.empty
+          else prev.failedScalafmtCheck & outDiff.unmodified
+        prevFailed foreach { warnBadFormat(_, log) }
+        val result = checkSources(filesToCheck.toSeq, config, log, writer)
+        prev.copy(
+          failedScalafmtCheck = result.failedScalafmtCheck | prevFailed
+        )
+    }
+  }
+
+  private def trueOrBoom(analysis: ScalafmtAnalysis): Boolean = {
+    val failureCount = analysis.failedScalafmtCheck.size
+    if (failureCount > 0) {
+      throw new MessageOnlyException(
+        s"${failureCount} files must be formatted"
+      )
+    }
+    true
+  }
+
+  private def warnBadFormat(file: File, log: Logger): Unit = {
+    log.warn(s"${file.toString} isn't formatted properly!")
   }
 
   private def checkSources(
       sources: Seq[File],
       config: Path,
       log: Logger,
       writer: PrintWriter
-  ): Boolean = {
-    val unformattedCount = withFormattedSources(sources, config, log, writer)(
-      (file, e) => {
-        log.err(e.toString)
-        false
-      },
+  ): ScalafmtAnalysis = {
+    log.info(s"Checking ${sources.size} Scala sources...")
```

Comment:
It's a little thing, but there's been a changed files count check before:

```
if (changed.size > 0) {
  log.info(s"Checking ${changed.size} Scala sources...")
}
```

So in your version I have these messages:

```
sbt:monixKafka> kafka10/scalafmtCheckAll
[info] Checking 0 Scala sources...
[info] Checking 0 Scala sources...
[success] Total time: 0 s, completed Aug 26, 2019 6:25:58 PM
```

Can we bring `if` block back?

```
if (sources.nonEmpty) {
  log.info(s"Checking ${sources.size} Scala sources...")
}
```